File: perf/benchmarks/micro/string_search/string_search.go
// string_search — substring-search tight-loop micro-benchmark (Go).
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

func main() {
	repeats := int64(50_000)
	if len(os.Args) > 1 {
		v, _ := strconv.ParseInt(os.Args[1], 10, 64)
		repeats = v
	}
	var seed int64 = 1
	buf := make([]byte, 0, 10_000)
	for i := 0; i < 10_000; i++ {
		seed = (seed*1103515245 + 12345) & 0x7FFFFFFF
		buf = append(buf, byte('a')+byte(seed%8))
	}
	hay := string(buf)
	needle := "abcde"

	t0 := time.Now()
	var cs int64 = 0
	for r := int64(0); r < repeats; r++ {
		var count int64 = 0
		pos := 0
		for {
			i := strings.Index(hay[pos:], needle)
			if i < 0 {
				break
			}
			count++
			pos += i + 1
		}
		cs ^= (r+1)*1000003 + count
	}
	ms := float64(time.Since(t0).Nanoseconds()) / 1_000_000.0
	fmt.Printf("CHECKSUM:%d\n", cs)
	fmt.Printf("ELAPSED_MS:%.6f\n", ms)
}

File: perf/benchmarks/micro/string_search/string_search.rs
// string_search — substring-search tight-loop micro-benchmark (Rust).
use std::env;
use std::time::Instant;

fn main() {
    let args: Vec<String> = env::args().collect();
    let repeats: i64 = if args.len() > 1 { args[1].parse().unwrap_or(50_000) } else { 50_000 };

    let mut seed: i64 = 1;
    let mut hay = String::with_capacity(10_000);
    for _ in 0..10_000 {
        seed = (seed.wrapping_mul(1103515245).wrapping_add(12345)) & 0x7FFF_FFFF;
        hay.push((b'a' + (seed % 8) as u8) as char);
    }
    let needle = "abcde";

    let t0 = Instant::now();
    let mut cs: i64 = 0;
    for r in 0..repeats {
        let mut count: i64 = 0;
        let mut pos: usize = 0;
        while let Some(i) = hay[pos..].find(needle) {
            count += 1;
            pos += i + 1;
        }
        cs ^= (r + 1) * 1_000_003 + count;
    }
    let ms = t0.elapsed().as_nanos() as f64 / 1_000_000.0;
    println!("CHECKSUM:{}", cs);
    println!("ELAPSED_MS:{:.6}", ms);
}
